src/PureScript/Backend/Optimizer/Convert.purs

```diff
@@ -232,14 +235,24 @@ toBackendTopLevelBindingGroup env = case _ of
   overValue f a =
     a { value = f a.value }
 
+-- | For the NonEmptyArray,
+-- | - `head` = the original expression
+-- | - `last` = the final optimized expression
+-- | - everything in-between the two are the steps that were taken from `head` to `last`
+type OptimizationSteps = Map Ident (NonEmptyArray BackendExpr)
+
 toTopLevelBackendBinding :: Array (Qualified Ident) -> ConvertEnv -> Binding Ann -> Accum ConvertEnv (Tuple Ident (WithDeps NeutralExpr))
 toTopLevelBackendBinding group env (Binding _ ident cfn) = do
   let evalEnv = Env { currentModule: env.currentModule, evalExtern: makeExternEval env, locals: [], directives: env.directives }
+  let qualifiedIdent = Qualified (Just env.currentModule) ident
   let backendExpr = toBackendExpr cfn env
-  let Tuple impl expr' = toExternImpl env group (optimize (getCtx env) evalEnv (Qualified (Just env.currentModule) ident) env.rewriteLimit backendExpr)
+  let enableTracing = env.traceOptimization env.currentModule ident
+  let Tuple mbSteps optimizedExpr = optimize enableTracing (getCtx env) evalEnv (Qualified (Just env.currentModule) ident) env.rewriteLimit backendExpr
+  let Tuple impl expr' = toExternImpl env group optimizedExpr
```

Comment:
One thing I noticed here. Currently, we pass in `Qualified (Just env.currentModule) ident` to optimize. I'm curious why we don't pass both separately (e.g. `optimize .. .. .. env.currentModule ident .. ..` since the `Qualified` constructor's module name is always wrapped in `Just`. That would simplify the `optimizeStep` crash case slightly.